builder: introduce chunkdict save subcommand

cslinwang opened the pull request:
## Relevant Issue (if applicable)

_N/A_

## Details

These commits introduce the following changes:

Background: In high-density deployment scenarios, the issue of redundant storage and distribution of chunks in Nydus images poses significant challenges. This problem leads to inefficient utilization of storage space and excessive consumption of bandwidth resources. To address this problem, it is crucial to design an effective chunk deduplication mechanism. This entails conducting in-depth research on chunk deduplication algorithms and developing solutions that enable deduplication across different business contexts, within the same business, across different layers, and within the same layer. By implementing comprehensive chunk deduplication strategies, we can significantly optimize storage utilization and enhance bandwidth efficiency in high-density deployment environments.

1. Commit: "cargo: Add rusqlite package to dependencies"
    
    - Adds the 'rusqlite' package to the project's dependencies.
    - Updates the 'cargo.toml' file to include the 'rusqlite' package.
    - Enables interaction with SQLite databases in the project.
2. Commit: "nydus-image: Store chunk and blob metadata"
    
    - Adds functionality to store chunk and blob metadata from nydus source images.
    - Introduces the 'nydus-image chunkdict save' command.
    - Allows storing metadata by using the '--bootstrap' option followed by the path to the nydus bootstrap file (e.g., '~/output/nydus_bootstrap').

## Types of changes

What types of changes do your commits introduce? Put an `x` in all the boxes that apply:

- [ ]  Bug fix (non-breaking change which fixes an issue)
- [x]  New feature (non-breaking change which adds functionality)
- [ ]  Breaking change (fix or feature that would cause existing functionality to change)
- [ ]  Documentation Update (if none of the other choices apply)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply:

- [x]  I have updated the documentation accordingly.
- [ ]  I have added tests to cover my changes.

Desiki-high commented:
/retest

adamqqqplay commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,239 @@
+            WHERE NOT EXISTS (
+                SELECT blob_id
+                FROM blob
+                WHERE blob_id = ?4
```
Is it possible to use `WHERE blob_id = ?1` here?

cslinwang replied:
Modified.

adam3q reviewed (commented):
Thanks for your contribution. This feature looks interesting, but maybe it could be more closely integrated with previous work #956 ? 

I've made some suggestions to your code. Also, some unit tests need to be added and the docs/nydus-image.md document updated. Thanks in advance.

cslinwang commented on Cargo.toml:
```diff
@@ -58,13 +59,17 @@ openssl = { version = "0.10.48", features = ["vendored"] }
 nydus-api = { version = "0.2.2", path = "api", features = ["handler"] }
 nydus-rafs = { version = "0.2.2", path = "rafs", features = ["builder"] }
-nydus-service = { version = "0.2.0", path = "service", features = ["block-device"] }
+nydus-service = { version = "0.2.0", path = "service", features = [
```
Sure, I apologize for overlooking it earlier.

imeoer replied:
Any updates?

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,239 @@
+        let tree = Tree::from_bootstrap(&self.sb, &mut ()).context(err)?;
+
+        // Connect to the database
+        let database_file: &str = "metadata.db";
```
We need to support various database protocols (not only for sqlite), so I think a abstract `Database` trait should be introduced, like:

```
trait Database {
    fn insert_chunk(&mut self, chunk: &Chunk) -> Result<()>;
    fn insert_blob(&mut self, blob: &Blob) -> Result<()>;
    fn get_chunk(&self) -> Result<Chunk>;
    fn get_blob(&self) -> Result<Blob>;
}
```

imeoer replied:
And please also add some unit test cases.

cslinwang replied:
Ok.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,239 @@
+    fn clear(conn: Connection) -> Result<()>;
+
+    //crate table
+    fn creat_table(conn: &Connection) -> Result<()>;
```
```
// create table
fn create(conn: &Connection) -> Result<()>;
```

cslinwang replied:
Refactored

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,239 @@
+
+    /// save metadata to database: chunk and blob info
+    pub fn save_metadata_to_database(&mut self) -> Result<Vec<Arc<BlobInfo>>> {
+        let err = "failed to load bootstrap for deduplicate";
```
Why define a variable for one-time use?

cslinwang replied:
Got it. Modified.

cslinwang commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,239 @@
+// Copyright 2020 Ant Group. All rights reserved.
```
Modified.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,239 @@
+}
+
+#[derive(Debug)]
+#[allow(dead_code)]
```
We should not allow dead codes.

cslinwang replied:
Ok.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,239 @@
+        conn.execute(
+            "CREATE TABLE IF NOT EXISTS blob (
+                blob_id             TEXT PRIMARY KEY ,
+                blob_size           TEXT NOT NULL,
```
Why not use INT type?

cslinwang replied:
Modified.

cslinwang commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,239 @@
+        let tree = Tree::from_bootstrap(&self.sb, &mut ()).context(err)?;
+
+        // Connect to the database
+        let database_file: &str = "metadata.db";
```
You are correct. I have added an additional parameter, "--database," to allow users to specify the storage location. I have also implemented handling logic for existing files.

cslinwang commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,239 @@
+
+        // Connect to the database
+        let database_file: &str = "metadata.db";
+        let conn = Connection::open(database_file)?;
```
Connection will be automatically closed when it goes out of scope.

cslinwang commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,239 @@
+                &BlobTable {
+                    blob_id: blob.blob_id().to_string(),
+                    blob_size: blob.uncompressed_size().to_string(),
+                    image_name: format_blob_features(blob.features()),
```
Modified.

cslinwang commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,239 @@
+                },
+            )?;
+        }
+        let pre = &mut |t: &Tree| -> Result<()> {
```
Modified.

imeoer commented:
cc @mofishzz Please also help to take a look, thanks!

cslinwang commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,239 @@
+                ChunkTable::insert_data_into_table(
+                    &conn,
+                    &ChunkTable {
+                        id: 1,
```
Refactored

cslinwang commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,239 @@
+    fn creat_table(conn: &Connection) -> Result<()> {
+        conn.execute(
+            "CREATE TABLE IF NOT EXISTS chunk (
+                id               INTEGER PRIMARY KEY ,
```
Refactored

cslinwang commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,239 @@
+        }
+        Ok(blobs)
+    }
+}
```
Modified.

cslinwang commented on src/bin/nydus-image/main.rs:
```diff
@@ -1038,6 +1088,36 @@ impl Command {
+        let blobs = deduplicate
+            .save_metadata_to_database()
+            .with_context(|| format!("failed to check bootstrap {:?}", bootstrap_path))?;
+        println!("RAFS filesystem metadata is saved.");
```
Modified.

cslinwang commented on src/bin/nydus-image/main.rs:
```diff
@@ -348,6 +350,47 @@ fn prepare_cmd_args(bti_string: &'static str) -> App {
+                        Arg::new("bootstrap")
+                            .short('B')
+                            .long("bootstrap")
+                            .help("[Deprecated] File path of RAFS meta blob/bootstrap")
```
Modified.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,361 @@
+// 2022 Nydus Developers.
```
2023

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,361 @@
+use std::error::Error;
+use std::fs;
+use std::path::Path;
+use std::sync::{Arc, Mutex};
```
Give a blank line here.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,361 @@
+use std::path::Path;
+use std::sync::{Arc, Mutex};
+pub trait Database {
+    fn create_chunk(&self) -> Result<(), Box<dyn Error + Send + Sync>>;
```
Can we give some comments for these methods?

cslinwang replied:
Ok.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,361 @@
+}
+
+impl Database for SqliteDatabase {
+    fn create_chunk(&self) -> Result<(), Box<dyn Error + Send + Sync>> {
```
Can we use `anyhow::Error` as the default Error type?

cslinwang replied:
Already modified.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,361 @@
+use std::sync::{Arc, Mutex};
+pub trait Database {
+    fn create_chunk(&self) -> Result<(), Box<dyn Error + Send + Sync>>;
+    fn create_blob(&self) -> Result<(), Box<dyn Error + Send + Sync>>;
```
Why use `Box<dyn Error + Send + Sync>`? How about `anyhow::Result<()>`?

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,361 @@
+        if let Ok(metadata) = fs::metadata(database_path) {
+            if metadata.is_file() {
+                if let Err(err) = fs::remove_file(database_path) {
+                    eprintln!(
```
`warn!`

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,361 @@
+    // clear table
+    fn clear(conn: &Connection) -> Result<(), rusqlite::Error>;
+
+    //create table
```
A space should be added, and use `///`.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,361 @@
+    //create table
+    fn create(conn: &Connection) -> Result<(), rusqlite::Error>;
+
+    //insert data
```
Ditto.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,361 @@
+    //insert data
+    fn insert(conn: &Connection, table: &Self) -> Result<(), rusqlite::Error>;
+
+    // select data
```
Should be `list` instead of `select`? How can we handle a large number of rows if no offset + limit?

cslinwang replied:
Ok.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,361 @@
+
+pub trait Table: Sync + Send + Sized + 'static {
+    // clear table
+    fn clear(conn: &Connection) -> Result<(), rusqlite::Error>;
```
Why use sqlite-related error type `rusqlite::Error` here?

cslinwang replied:
Modified to a generic type.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,361 @@
+    sb: RafsSuper,
+    db: D,
+}
+impl Deduplicate<SqliteDatabase> {
```
Give a space here.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,361 @@
+        Ok(Self { sb, db })
+    }
+
+    /// save metadata to database: chunk and blob info
```
The first letter should be capitalized.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,361 @@
+#[derive(Debug)]
+pub struct BlobTable {
+    blob_id: String,
+    blob_size: u64,
```
Add `blob_compressed_size` and `blob_uncompressed_size`?

cslinwang replied:
Ok.

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -348,6 +350,56 @@ fn prepare_cmd_args(bti_string: &'static str) -> App {
+                        Arg::new("database-type")
+                            .long("database-type")
+                            .help("The type of metadata database, currently supported is Sqlite.")
+                            .default_value("Sqlite")
```
`sqlite`.

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -348,6 +350,56 @@ fn prepare_cmd_args(bti_string: &'static str) -> App {
+                            .required(false),
+                        )
+                    .arg(
+                        Arg::new("database-path")
```
How about using a database URL like `sqlite:///path/to/image.db`?

cslinwang replied:
Ok.

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -1038,6 +1098,49 @@ impl Command {
+            }
+            // Add other cases for different database types if needed
+            _ => {
+                return Err(anyhow!("Unsupported database type"));
```
Should give the type name in the error message.

cslinwang replied:
Improved prompt messages。

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -1038,6 +1098,49 @@ impl Command {
+                return Err(anyhow!("Unsupported database type"));
+            }
+        };
+        println!("RAFS filesystem metadata is saved:");
```
`info!`

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -1038,6 +1098,49 @@ impl Command {
+        println!("RAFS filesystem metadata is saved:");
+        let mut blob_ids = Vec::new();
+        for (idx, blob) in blobs.iter().enumerate() {
+            println!(
```
`info!`

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,496 @@
+// 2023 Nydus Developers.
```
// Copyright (C) 2023 Nydus Developers. All rights reserved.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,496 @@
+
+pub trait Database {
+    /// Creates a new chunk in the database.
+    fn create_chunk(&self) -> Result<(), ()>;
```
Should be `create_chunk_table`?

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,496 @@
+    fn create_chunk(&self) -> Result<(), ()>;
+
+    /// Creates a new blob in the database.
+    fn create_blob(&self) -> Result<(), ()>;
```
Should be `create_blob_table`?

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,496 @@
+    pub fn new(
+        bootstrap_path: &Path,
+        config: Arc<ConfigV2>,
+        db_path: &str,
```
Should be `db_url`? It's not only for sqlite in the future.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,496 @@
+        db_path: &str,
+    ) -> anyhow::Result<Self> {
+        let (sb, _) = RafsSuper::load_from_file(bootstrap_path, config, false)?;
+        let db = SqliteDatabase::new(db_path.strip_prefix('/').unwrap_or(db_path))?;
```
Why need `strip_prefix` here? We'd better limit the database path to an absolute path.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,496 @@
+    }
+
+    /// Save metadata to the database: chunk and blob info.
+    pub fn save_metadata(&mut self, _mode: Option<&str>) -> anyhow::Result<Vec<Arc<BlobInfo>>> {
```
What is this `_mode` used for?

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,496 @@
+        // Create the blob table and chunk table.
+        self.db
+            .create_chunk()
+            .map_err(|e| anyhow!("Failed to create chunk: {:?}.", e))?;
```
Can we use `.context` here?

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,496 @@
+            .map_err(|e| anyhow!("Failed to create chunk: {:?}.", e))?;
+        self.db
+            .create_blob()
+            .map_err(|e| anyhow!("Failed to create blob: {:?}.", e))?;
```
Ditto.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,496 @@
+        }
+
+        // Save chunk info to the chunk table.
+        let pre = &mut |t: &Tree| -> anyhow::Result<()> {
```
`anyhow::Result<()>` -> `Result<()>`.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,496 @@
+        let pre = &mut |t: &Tree| -> anyhow::Result<()> {
+            let node = t.lock_node();
+            for chunk in &node.chunks {
+                let index: u32 = chunk.inner.blob_index();
```
Why need `u32` type hint here?

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,496 @@
+                        chunk_compressed_offset: chunk.inner.compressed_offset(),
+                        chunk_uncompressed_offset: chunk.inner.uncompressed_offset(),
+                    })
+                    .map_err(|e| anyhow!("Failed to insert chunk: {:?}.", e))?;
```
Ditto.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,496 @@
+    Err: Error + 'static,
+{
+    /// clear table.
+    fn clear(conn: &Conn) -> Result<(), Err>;
```
It feels a little weird to pass conn in every method, can we put `conn` into `ChunkTable`, `BlobTable` as a struct field?

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,496 @@
+}
+
+#[derive(Debug)]
+pub struct ChunkTable {
```
How about:

```
struct ChunkTable {
    conn: Conn,
}

struct Chunk {
    chunk_blob_id: String,
    chunk_digest: String,
    chunk_compressed_size: u32,
    chunk_uncompressed_size: u32,
    chunk_compressed_offset: u64,
    chunk_uncompressed_offset: u64,
}
```

cslinwang replied:
We cannot make the Chunk and ChunkTable structs private. This is because the implementation of Deduplicate is designed to ensure the diversity of database types, specified as: pub struct Deduplicate<D: Database + Send + Sync>. Additionally, the methods within Deduplicate are interrelated with each other.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,496 @@
+            "CREATE TABLE IF NOT EXISTS chunk (
+                id               INTEGER PRIMARY KEY,
+                chunk_blob_id    TEXT NOT NULL,
+                chunk_digest     TEXT,
```
Should be `TEXT NOT NULL`?

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,496 @@
+
+    fn list_all(conn: &rusqlite::Connection) -> Result<Vec<Self>, rusqlite::Error> {
+        let mut offset = 0;
+        let limit: i64 = 100; // 每页的行数
```
Use english comment here.

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -348,6 +350,49 @@ fn prepare_cmd_args(bti_string: &'static str) -> App {
+                    .arg(
+                        Arg::new("database-path")
+                            .long("database-path")
+                            .help("File path of metadata database, like: 'sqlite:///path/to/database.db'")
```
Database connection URI for assisting chunk dict generation, e.g. `sqlite:///path/to/database.db`.

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -348,6 +350,49 @@ fn prepare_cmd_args(bti_string: &'static str) -> App {
+                        Arg::new("database-path")
+                            .long("database-path")
+                            .help("File path of metadata database, like: 'sqlite:///path/to/database.db'")
+                            .default_value("sqlite:///metadata.db")
```
Should be no default value for this option.

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -348,6 +350,49 @@ fn prepare_cmd_args(bti_string: &'static str) -> App {
+                        )
+                    .arg(
+                        Arg::new("database-path")
+                            .long("database-path")
```
`database` is ok.

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -1038,6 +1091,53 @@ impl Command {
+        let db_type: Vec<&str> = db_path.split("://").collect();
+
+        if db_type.is_empty() {
+            return Err(anyhow!("Invalid or unsupported database path. Please provide a correct database path, such as sqlite:///path/to/database.db"));
```
`bail!("Invalid database URL")`

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -1038,6 +1091,53 @@ impl Command {
+            return Err(anyhow!("Invalid or unsupported database path. Please provide a correct database path, such as sqlite:///path/to/database.db"));
+        }
+
+        let blobs: Vec<Arc<nydus_storage::device::BlobInfo>> = match db_type[0] {
```
`use nydus_storage::device::BlobInfo`.

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -1038,6 +1091,53 @@ impl Command {
+            .internal
+            .set_blob_accessible(matches.get_one::<String>("bootstrap").is_none());
+
+        let db_type: Vec<&str> = db_path.split("://").collect();
```
`db_strs`

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -1038,6 +1091,53 @@ impl Command {
+
+        let db_type: Vec<&str> = db_path.split("://").collect();
+
+        if db_type.is_empty() {
```
Should be `if db_strs.len() != 2 {`

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -1038,6 +1091,53 @@ impl Command {
+                    Deduplicate::<SqliteDatabase>::new(bootstrap_path, config, db_type[1])?;
+                deduplicate
+                    .save_metadata(Some(&db_path))
+                    .with_context(|| format!("failed to check bootstrap {:?}.", bootstrap_path))?
```
Why the error is related with `check bootstrap`?

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -1038,6 +1091,53 @@ impl Command {
+                    .with_context(|| format!("failed to check bootstrap {:?}.", bootstrap_path))?
+            }
+            _ => {
+                return Err(anyhow!("Unsupported database type: {}. Please provide a correct database type, such as sqlite.", db_type[0]));
```
`bail!("Unsupported database type: {}, please use a valid database URI, such as 'sqlite:///path/to/database.db'.")`

imeoer commented:
@cslinwang It seems to have an abnormal commit `Merge branch 'master' into linwang-dev`.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,661 @@
+        let chunk_table = match chunk_table_result {
+            Ok(table) => table,
+            Err(e) => {
+                println!("An error occurred when creating the ChunkTable: {}", e);
```
Should use `error!`.

cslinwang replied:
Resolved

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,661 @@
+        let blob_table = match blob_table_result {
+            Ok(table) => table,
+            Err(e) => {
+                println!("An error occurred when creating the BlobTable: {}", e);
```
Ditto.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,661 @@
+        let chunk_table = match chunk_table_result {
+            Ok(table) => table,
+            Err(e) => {
+                println!("An error occurred when creating the ChunkTable: {}", e);
```
Ditto.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,661 @@
+        let blob_table = match blob_table_result {
+            Ok(table) => table,
+            Err(e) => {
+                println!("An error occurred when creating the BlobTable: {}", e);
```
Ditto.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,661 @@
+}
+
+pub struct SqliteDatabase {
+    chunk_table: Option<ChunkTable>,
```
Why do we need to wrap into `Option` here?

cslinwang replied:
Historical legacy issue resolved.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,661 @@
+        let blob_table = BlobTable::new_in_memory()?;
+        {
+            let conn = blob_table.conn.lock().map_err(|_| "Mutex poisoned")?;
+            conn.execute(
```
Why not use `Table::create` here.

imeoer commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,661 @@
+        {
+            let conn = chunk_table.conn.lock().map_err(|_| "Mutex poisoned")?;
+            conn.execute(
+                "CREATE TABLE IF NOT EXISTS chunk (
```
Ditto.

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -356,6 +358,49 @@ fn prepare_cmd_args(bti_string: &'static str) -> App {
+                        Arg::new("database")
+                            .long("database")
+                            .help("Database connection URI for assisting chunk dict generation, e.g. sqlite:///path/to/database.db.")
+                            .default_value("sqlite:///sqlite_memory_model")
```
How about using `sqlite://:memory:`, reference: https://www.sqlite.org/inmemorydb.html

imeoer commented on Cargo.toml:
```diff
@@ -46,6 +46,7 @@ log-panics = { version = "2.1.0", features = ["with-backtrace"] }
 mio = { version = "0.8", features = ["os-poll", "os-ext"] }
 nix = "0.24.0"
 rlimit = "0.9.0"
+rusqlite = { version = "0.29.0", features = ["bundled"] }
```
Does Cargo.lock need to be updated?

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -356,6 +358,49 @@ fn prepare_cmd_args(bti_string: &'static str) -> App {
+                    .arg(
+                        Arg::new("database")
+                            .long("database")
+                            .help("Database connection URI for assisting chunk dict generation, e.g. sqlite:///path/to/database.db.")
```
Remove the period of tail.

cslinwang committed:
cargo: Add rusqlite package to dependencies

Updates 'Cargo.toml' and 'Cargo.lock' files to include the 'rusqlite' package.
Enabling interaction with SQLite databases in the project.

Signed-off-by: Lin Wang <l.wang@mail.dlut.edu.cn>

cslinwang committed:
nydus-image: Store chunk and blob metadata

Add functionality to store chunk and blob metadata
from nydus source images.
Use the 'nydus-image chunkdict save' command
with the '--bootstrap' option followed by the
path to the nydus bootstrap file (e.g., '~/output/nydus_bootstrap')

Signed-off-by: Lin Wang <l.wang@mail.dlut.edu.cn>

imeoer reviewed (approved):
LGTM!
